documentation/adr/0023-error-format.md

```diff
@@ -0,0 +1,88 @@
+herbert_core/utilities/xml_io_tools/xmlwrite_xerces.m                                         : xml:FileNotFound
+herbert_core/admin/matlab_nbits.m                                                             : MATLAB:NBITS
+
+_test/shared/matlab_xunit_ISISextras/@TestCaseWithSave/private/get_ref_dataset_.m             : TestCaseWithSave:invalid_argument
+_test/shared/matlab_xunit_ISISextras/@TestCaseWithSave/private/instantiate_methods_to_save_.m : TEST_CASE_WITH_SAVE:invalid_argument
+
+erbert_core/classes/MPIFramework/@iMessagesFramework/iMessagesFramework.m                     : iMESSAGES_FRAMEWORK:invalid_argument
+herbert_core/classes/MPIFramework/@iMessagesFramework/iMessagesFramework.m                    : MESSAGES_FRAMEWORK:invalid_argument
+herbert_core/classes/MPIFramework/@iMessagesFramework/private/mix_messages_.m                 : iMESSAGES_FRAMEWOR:invalid_argument
+```
+From which it should be obvious that this is not a consistent scheme for error identification, even within the same set of routines.
+
+As MATLAB standards do not enforce a particular style of error statement 
+(though it does encourage certain forms [see here](https://uk.mathworks.com/help/matlab/ref/mexception.html#mw_e5712c7f-3862-42fa-9a8f-8de992cdc6d4)),
+this document seeks to outline the format an error identifier should take as well as establish a list of standard common error identifiers.
+
+Also due to the fact that MATLAB has changed its method of error identification between versions in the past, this document establishes its scheme based on the most recent currently used version (2020b).
+
+## Decision
+
+Following discussions it was decided that it would be useful to have a standard format with 3 components, in the following format
+(expanding upon that intially outlined in [ADR 21](./0021-errors-and-warnings.md)).
+
+
+```matlab
+error('(HORACE|HERBERT):(function_name|ClassName):error_identifier')
+```
+
+This scheme is based on the current (2020b) MATLAB error identifiers common to many functions.
+
+- `HORACE` or `HERBERT` should be in all caps.
+- The `function_name` and `ClassName` should exactly match that of the parent function/class, including casing.
+- The `error_identifier` should be in `lower_snake` case and should succinctly identify the error message, using one of
+     the following names if the error falls within their remit.
+     **NB.** This is in contrast to the usual MATLAB format which uses `lowerCamelCase`, but is in keeping with the majority
+     of identifiers currently used in PACE.
+
+| Identifier         | Issue                                                                                         |
+| :----------------- | :-------------------------------------------------------------------------------------------- |
+| `invalid_argument` | Argument fails validation, there are insufficient arguments or argument is an unexpected flag |
```

Comment:
What happens if we're using something like `validattributes`? Should we be catching the errors and re-throwing with our ID? I'm not sure that I mind either way, but a convention on what to do in these cases would be good.

Reply:
A very good question. I would probably say use MATLAB's throw. The advantage of throwing our own means we know exactly what it is and it is consistent. 

Eventually (you might say I'm a dreamer, but I'm not the only one), I can see either MATLAB's `inputParser`/`validateattributes` or PACE's `parse_args` being used globally and, either way, consistency will be achieved. Even if we end up slowly phasing `invalid_argument` or MATLAB's equivalent, that's really the key. 

`invalid_argument` will be useful for those times where parsing beyond that of validate is used (though admittedly if `inputParser` is used that argument becomes moot with the ability to define arbitrary validation.